docs/core/tools/dotnet-environment-variables.md

```diff
@@ -183,6 +183,22 @@ See [EventPipe environment variables](../diagnostics/eventpipe.md#trace-using-en
 
 Specifies the location of the .NET runtimes, if they are not installed in the default location. The default location on Windows is `C:\Program Files\dotnet`. The default location on Linux and macOS is `/usr/local/share/dotnet`. This environment variable is used only when running apps via generated executables (apphosts). `DOTNET_ROOT(x86)` is used instead when running a 32-bit executable on a 64-bit OS.
 
+### `DOTNET_HOST_PATH`
+
+Specifies the absolute path to a `dotnet` binary (`dotnet.exe` on Windows, `dotnet` on non-Windows) that was used to launch this session. This is typically used by a hosting application that wants to ensure that the exact same version of the host binary is used for all child processes.
+
+Without this environment variable, if a `dotnet` binary was used from a location not on the `PATH`, child processes of that `dotnet` that wished to invoke `dotnet` would invoke a different `dotnet` binary than the one that was used to spawn them - this can result in confusing behavior for end users.
+
+As a tool author, if you want to locate the correct `dotnet` binary to use to spawn a .NET binary, you should use the following algorithm:
+
+* if `DOTNET_HOST_PATH` is set, use that value directly
+* otherwise, probe `PATH` according to the rules of your operating system to find the first `dotnet` binary on the `PATH`
```

Comment:
I wouldn't recommend probing `PATH` according to the rules of your OS. Instead I would suggest to just launch `dotnet` (or `dotnet.exe` depending on OS) and let the OS take care of locating the executable.

cc @vitek-karas

Reply:
That only works if you're using `UseShellExecute=true` though, if I'm remembering right - and many folks spawn commands and redirect the stdin/stdout/stderr which blocks that execution pathway.

Reply:
In that case we should pull in security folks - I remember that parsing `PATH` is considered a bad idea, especially on Windows.

Otherwise I completely agree with Eric that we should let the OS sort it out.

Reply:
At an MSBuild task level you can also just set the tool exe to `dotnet` and let MSBuild probe `$PATH` for you (that is how it works today).

Reply:
This document should not be discussing `PATH` at all. It should describe how this ENV is intended to be used, when it is typically set by others, and if/when you (the user) should set it.

Reply:
> I think we need to discuss what to do when this value is not set.

My recommendation/thought/opinion is above. Copied here:

> Instead I would suggest to just launch dotnet (or dotnet.exe depending on OS) and let the OS take care of locating the executable.

Reply:
Okay but how does one do that in the context of an MSBuild task? What specifically do we want authors to do here? Cause in the case of `ToolTask` there isn't an option of "let OS take care of it". 

I know that as an implementation detail, you can achieve this by providing the tool path to be simply `dotnet` and then MSBuild will do the work for you. That is not documented AFAIK though.

Reply:
I believe https://github.com/dotnet/msbuild/issues/9018 is an attempt to try to remedy this for MSBuild. Honestly, I have no idea why MSBuild parses `$PATH` itself, I'm sure it was to work around some corner-case issue. It feels unnecessary. cc @rainersigwald. 

> Okay but how does one do that in the context of an MSBuild task? What specifically do we want authors to do here?

Just say `dotnet` or `dotnet.exe`, as necessary.

Reply:
> Just say dotnet or dotnet.exe, as necessary.

Then we need to put this in guidance, not ignore it.

Reply:
> Just say `dotnet` or `dotnet.exe`, as necessary.

Is that . . . ok? Wouldn't it run different bits in some circumstances, like using a specific old SDK while a newer SDK/runtime is installed?

Reply:
> Wouldn't it run different bits in some circumstances, like using a specific old SDK while a newer SDK/runtime is installed?

Can you describe the situation you are thinking about more? I'm not following what you'd think would be a problem here, over parsing `$PATH` yourself.

Note if you are already inside a .NET SDK command (like `dotnet build`), then `DOTNET_HOST_PATH` will already be set, so you won't fall into this case where `DOTNET_HOST_PATH` isn't set.

Reply:
You're right, I think that completely avoids my concern. In that case we can simplify the MSBuild helper to "value of `DOTNET_HOST_PATH` or `dotnet`/`dotnet.exe`".

Reply:
MSBuild has an issue to investigate this now, and the post has been updated.